source/Calamari/Kubernetes/Commands/KubernetesDiscoveryCommand.cs

```diff
@@ -0,0 +1,168 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Calamari.Commands.Support;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Discovery;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.ServiceMessages;
+using Calamari.Common.Plumbing.Variables;
+using Calamari.Kubernetes.Commands.Discovery;
+using Newtonsoft.Json;
+
+namespace Calamari.Kubernetes.Commands
+{
+    [Command(Name, Description = "Discovery Kubernetes cluster targets")]
+    public class KubernetesDiscoveryCommand : Command
+    {
+        public const string Name = "kubernetes-target-discovery";
+        public const string CreateKubernetesTargetServiceMessageName = "create-kubernetestarget";
+        public const string ContextVariableName = "Octopus.TargetDiscovery.Context";
+        
+        readonly ILog log;
+        readonly IVariables variables;
+        readonly IKubernetesDiscovererFactory discovererFactory;
+
+        public KubernetesDiscoveryCommand(ILog log,
+            IVariables variables,
+            IKubernetesDiscovererFactory discovererFactory)
+        {
+            this.log = log;
+            this.variables = variables;
+            this.discovererFactory = discovererFactory;
+        }
+
+        /// <returns>
+        /// The Discovery Command always returns 0 to indicate success
+        /// because a failure to discover targets should not cause a
+        /// deployment process to fail. The one exception is if the
+        /// json is malformed as this can only occur if there is
+        /// a code issue and not from user misconfiguration.
+        /// </returns>
+        public override int Execute(string[] commandLineArguments)
+        {
+            if (!TryGetDiscoveryContextJson(out var json))
+            {
+                log.Warn($"Could not find target discovery context in variable {ContextVariableName}.");
+                return 0;
+            }
+            
+            if (!TryGetAuthenticationContextTypeAndDiscoveryContextScope(json, out var type, out var scope))
+                return -1;
```

Comment:
I am curious is this the right return code? I don't know if we interpret these error codes in any way in the server code.

Reply:
Sorry, that was my placeholder - I've updated this now to match what other random exceptions would return.